Dockerfile

```diff
@@ -1,45 +1,86 @@
+    rm -f /etc/apt/apt.conf.d/docker-clean
+    echo 'Binary::apt::APT::Keep-Downloaded-Packages "true";' > /etc/apt/apt.conf.d/keep-cache
+
+    apt update && apt install -y --no-install-recommends \
+      build-essential \
+      curl \
+      python3-venv \
+      cmake
+HEREDOC
+
 ENV HOME="/root"
+ENV PATH="$HOME/.venv/bin:$PATH"
 WORKDIR $HOME
 
-RUN apt update \
-  && apt install -y --no-install-recommends \
-  build-essential \
-  curl \
-  python3-venv \
-  cmake \
-  && apt clean \
-  && rm -rf /var/lib/apt/lists/*
-
-# Setup zig as cross compiling linker
-RUN python3 -m venv $HOME/.venv
-RUN .venv/bin/pip install cargo-zigbuild
-ENV PATH="$HOME/.venv/bin:$PATH"
+# Setup zig as cross compiling linker:
+RUN <<HEREDOC
+  python3 -m venv $HOME/.venv
+  .venv/bin/pip install --no-cache-dir cargo-zigbuild
+HEREDOC
 
-# Install rust
-ARG TARGETPLATFORM
-RUN case "$TARGETPLATFORM" in \
-  "linux/arm64") echo "aarch64-unknown-linux-musl" > rust_target.txt ;; \
-  "linux/amd64") echo "x86_64-unknown-linux-musl" > rust_target.txt ;; \
-  *) exit 1 ;; \
-  esac
-# Update rustup whenever we bump the rust version
-COPY rust-toolchain.toml rust-toolchain.toml
-RUN curl --proto '=https' --tlsv1.2 -sSf https://sh.rustup.rs | sh -s -- -y --target $(cat rust_target.txt) --profile minimal --default-toolchain none
+# Install rust:
 ENV PATH="$HOME/.cargo/bin:$PATH"
-# Installs the correct toolchain version from rust-toolchain.toml and then the musl target
-RUN rustup target add $(cat rust_target.txt)
-
-# Build
-COPY crates crates
-COPY ./Cargo.toml Cargo.toml
-COPY ./Cargo.lock Cargo.lock
-RUN cargo zigbuild --bin uv --target $(cat rust_target.txt) --release
-RUN cp target/$(cat rust_target.txt)/release/uv /uv
-# TODO(konsti): Optimize binary size, with a version that also works when cross compiling
-# RUN strip --strip-all /uv
+COPY rust-toolchain.toml .
+RUN <<HEREDOC
+  # Install rustup, but skip installing a default toolchain as we only want the version from `rust-toolchain.toml`:
+  curl --proto '=https' --tlsv1.2 -sSf https://sh.rustup.rs | sh -s -- -y --profile minimal --default-toolchain none
+
+  # When rustup installs the toolchain ensure it actually uses the minimal profile, avoiding excess layer weight:
+  # https://github.com/rust-lang/rustup/issues/3805#issuecomment-2094066914
+  echo 'profile = "minimal"' >> rust-toolchain.toml
+  echo 'targets = [ "aarch64-unknown-linux-musl", "x86_64-unknown-linux-musl" ]' >> rust-toolchain.toml
+  # Add the relevant musl target triples (for a building binary with static linking):
+  # Workaround until `ensure` arrives: https://github.com/rust-lang/rustup/issues/2686#issuecomment-788825744
+  rustup show
+HEREDOC
+
+# Handle individual images differences for ARM64 / AMD64:
+FROM builder-base AS builder-arm64
+ENV CARGO_BUILD_TARGET=aarch64-unknown-linux-musl
+
+FROM builder-base AS builder-amd64
+ENV CARGO_BUILD_TARGET=x86_64-unknown-linux-musl
```

Comment:
This is effectively what you were doing earlier with the bash conditional statement on `TARGETPLATFORM` arg. But instead of writing to a file (`rust_target.txt`), a new intermediary stage is introduced and sets the expected target as an appropriate ENV.

This actually removes the need for the `--target` option when building, but I kept it for clarity.

The next stage (`builder-app`) refers to the `TARGETARCH` implicit arg from BuildKit. Thus when building the `Dockerfile` the stage selection here is determined from the final stage being built.

It'll remain as being built from the same `BUILDPLATFORM` above, but diverge at this point. Due to the cache mounts used in `builder-app`, this isn't a big concern, you really only have a few MB repeated from the `COPY` instruction, followed by the binary `cp` (_since the `target/` cache mount excludes the original binary from the image_).

A prior commit in this PR history had an alternative approach where both targets were built, and these separate stages were located at the end of the `Dockerfile`, where they instead used `COPY --from` with the location to their respective target binary.